Validate HTTP response codes across the extractor [1]

absurdlylongusername opened the pull request:
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [x] I have tested the API against [NewPipe](https://github.com/TeamNewPipe/NewPipe).
- [x] I agree to create a pull request for [NewPipe](https://github.com/TeamNewPipe/NewPipe) as soon as possible to make it compatible with the changed API.

### Initial PR to introduce validation of HTTP response codes for network requests throughout the extractor

(Part of #1472)

Almost all code that uses the Downloader does not do any checking on the response code: therefore errors and exceptions thrown due to error response code give false information (e.g. tests failing due to 404 were giving error about `clientId` when the error was 404).

TL;DR this PR adds methods in Response.java for validating response codes and uses it in some places for soundcloud extraction. These methods throw exceptions with information about the error code, and propagate upwards to the call site to print a more useful stack trace for debugging.

Read discussions below for full context.

I don't like chonky PRs so I will not do too much refactoring in this PR, but will split it up into multiple PRs

TobiGr reviewed (commented):
Thaank you. Changes LGTM, but conflicts need to be resolved first.

litetex reviewed (dismissed):
I don't think this can be merged in the current state:
* There are Todo comments
* The changes in the test will be irrelevant after #1332 
* Response codes that are ok are 2XX response codes
* I'm really not sure about another HttpUtils method and a new Exception class

I think the most simple approach would be to have a ``boolean Response#isOk()`` method and then use this method to control the underlying use cases accordingly.

davidasunmo commented:
@litetex 

> - There are Todo comments

The comment was kind of intentionally left in there in the hopes that whoever reviewed it would provide an explanation and then I would remove it.

> - Response codes that are ok are 2XX response codes

Not exactly sure what you mean by this? Do you mean that 2xx is the only "success" code range all network calls throughout the codebase?

>  - I'm really not sure about **another** HttpUtils method and a new Exception class
> I think the most simple approach would be to have a boolean Response#isOk() method and then use this method to control the underlying use cases accordingly.

Do you mean another as in having two `validateResponseCode` methods, one in `Response` and one in `HttpUtils`, or do you mean that there's another `HttpUtils` class somewhere (because I can't seem to find one)?

And does this also mean we shouldn't have a `HttpResponseException`? If so, why? I think having a specific exception for Http errors is better than throwing a generic `IOException`, and also makes for better logging and diagnostics, because right now our code does not give any useful information at all for what causes an error if the cause is an invalid response.

* * *
My idea _was_ to have a generic method we can use to check for expected response codes instead of just checking it's 2xx or not 2xx, because I was of the assumption that:

1. We are definitely not validating response codes anywhere in the codebase (I can't see anywhere we do this)
2. There may be scenarios where we want to treat a specific 3xx or 4xx as not an error and it's not efficient to catch an exception in these cases and _then_ check the response code (and then throw _another_ exception with info about the http error that will get propagated upwards), as I believe we should never throw exception if we don't have to, and exceptions should only be thrown for unexpected behaviour, or in the same scenarios we already throw exceptions.


The end goal is that **if a network call ever fails, we want to know why**, using code that is:

- As small as possible
- Easy to understand
- Easy to use
- Easy to extend

If there's only one place in code where non-2xx codes shouldn't be considered an error then we could handle that case explicitly at that call site, but if it's at least 3 times then I think we should have more than a simple `isOk` method. 
Could perhaps have other methods like Response#isRedirect for 3xx, Response#isClientError for 4xx, Response#isServerError instead of a single generic method.

Since I am almost certain 1. is true, I am more than happy to trawl through the entire codebase for every http call we make (not jusr for `get`, but for `post` and `head` too) and add response code validation and see which approach makes the most sense.

I could update this PR with these changes and/or split it into multiple PRs if it is too big.

Thoughts?

Stypox commented on extractor/src/main/java/org/schabi/newpipe/extractor/utils/HttpUtils.java:
```diff
@@ -0,0 +1,36 @@
+     * or when {@code  validResponseCodes} is empty and the code is a 4xx or 5xx error.
+     */
+    // CHECKSTYLE:ON
+    public static void validateResponseCode(final Response response,
```
I would inline this directly in `Response`, especially because it's first argument is a `Response`

Stypox commented on extractor/src/test/java/org/schabi/newpipe/extractor/services/soundcloud/SoundcloudChannelTabExtractorTest.java:
Yeah changes in this file should be reverted as they were superseded

Stypox commented on extractor/src/main/java/org/schabi/newpipe/extractor/exceptions/HttpResponseException.java:
```diff
@@ -0,0 +1,15 @@
+import java.io.IOException;
+import org.schabi.newpipe.extractor.downloader.Response;
+
+public class HttpResponseException extends IOException {
```
This should be an `ExtractorException`, since the extractor received an unexpected response. There was no error in data input/output, since the request went through just fine at the network layer.

Stypox reviewed (changes requested):
Thanks! We do need to check response codes more often. For example, PeerTube sometimes throws 429 Too Many Requests (https://github.com/TeamNewPipe/NewPipe/issues/2979), but that shouldn't be treated as a recaptcha by the `Downloader` (while that's what happens now). It should be the PeerTube service's job to throw a normal ExtractorException for 429s, and instead the YouTube service's job to throw a RecaptchaException (along with a proper URL that users can use to solve the captcha, instead of having the `Downloader` guess that URL).

It would be good if you could take the above paragraph into consideration for this PR, and properly validate the HTTP response codes of all requests made by all services in the extractor, so we don't have to change the design once again later.

absurdlylongusername commented on extractor/src/main/java/org/schabi/newpipe/extractor/downloader/Response.java:
```diff
@@ -80,4 +83,21 @@ public String getHeader(final String name) {
 
         return null;
     }
+    // CHECKSTYLE:OFF
```
IIRC, checkstyle was complaining about line being too long in the comment below, so I had to disable it

Stypox replied:
Not really important, but why don't you make the lines shorter then?

Stypox commented on extractor/src/main/java/org/schabi/newpipe/extractor/services/soundcloud/SoundcloudParsingHelper.java:
```diff
@@ -189,6 +194,7 @@ public static String resolveUrlWithEmbedPlayer(final String apiUrl) throws IOExc
      *
      * @return the resolved id
      */
+    // TODO: what makes this method different from the others? Don' they all return the same?
```
Which method would be the same as which others? They all do different things imo

Stypox commented:
> Now, what exactly is the difference between an ExtractionException and a ParsingException?

I guess this was a very early attempt to distinguish exceptions caused by getting and deserializing data from the internet, vs parsing that data using the methods defined in the extractor. However, I don't think that distinction holds now, and the there is no semantic distinction between the two. If you want to open a PR to delete `ParsingException` and replace it with `ExtractionException` everywhere feel free to.

> If we add http response validation throughout the extractor, should the HttpResponseException be swallowed and wrapped in a ParsingException? Because as far as I can see that is what happens throughout the codebase regardless of the type of exception thrown when we make network requests.

I would create a `NetworkException extends ExtractionException` to wrap all network-related exception. Error handling is generally difficult to do, but I think the current state of the extractor mostly works as expected: wrapping exceptions means we still get to look at the original exception in the stacktrace in bugreports, but the API remains clean so NewPipeExtractor users don't have to `catch` tens of exception types.

absurdlylongusername committed:
[SoundCloud] Validate http response code in SoundcloudParsingHelper

absurdlylongusername committed:
Move validateResponseCode to Response
Make HttpResponseException extend ExtractionException
Remove HttpUtils

absurdlylongusername committed:
Fix checkstyle off

absurdlylongusername committed:
Remove validateResponseCodes
Add various helper methods for validating response codes instead

absurdlylongusername committed:
Add validation to all http calls in SoundcloudParsingHelper

absurdlylongusername committed:
Catch HttpResponseException in to fix SoundcloudChartsExtractorTest.testRelatedItems and testMoreRelatedItems

absurdlylongusername commented:
@Stypox @TobiGr I've addressed the outstanding code review comments.

I want to get this PR merged ASAP so #1325 can be merged. So in this PR I just want to do initial setup of helper methods for `Response.java`, and adding response code validation in `SoundcloudParsingHelper`.

I added a bunch of initial helper methods that came to mind. I forgot to add `ensureValidResponseCodes` and `throwIfInvalidResponseCodes` that take vararg array of codes. 
Idk if all of them will be necessary anyway: as I add validation thoroughout the codebase in later PRs I will add/remove methods as needed, so perhaps `ensureValidResponseCodes` and `throwIfInvalidResponseCodes` may never be needed at all.

I did not do bounds checking on `min/max` in the range validation methods (I forgot). I will add them in the next PR, but I can add them in this one if you think it's very important enough.

Things I will do in next PR:

- Add tests for `Response` and update tests for `SoundcloudParsingHelper` to test validation
- Remove `ParsingException` and replace with ExtractorException
- Make `HttpResponseException` extends `NetworkException` extends `ExtractorException`
- Add some more validation in places, and add/update tests accordingly


Things worthy of note:

- I have not updated the mocks. I don't really know how: I know how to run the recording downloader, but I don't get how it works, because when I ran it to update mocks for failing tests it looks like it deleted all generated mocks for all the tests I didn't run for that specific folder (soundcloudparsinghelper).
- What region should mocks be run on? E.g. if I use VPN location in Spain or Sweden or Singapore, does it matter as long as the tests pass?
- I have not tested these (recently pushed) changes in NewPipe yet. Likely, if I do proper static analysis I could conclude that it's not possible for the code to actually introduce any bugs since in every case it throws an exception for invalid response code, it would throw an exception regardless: but I cannot confirm this. Also I haven't done unit tests yet either so can't confirm it for that reason too.

TobiGr commented on extractor/src/main/java/org/schabi/newpipe/extractor/downloader/Response.java:
```diff
@@ -80,4 +84,129 @@ public String getHeader(final String name) {
+        final Function<Response, HttpResponseException> errorSupplier
+    )
+        throws HttpResponseException {
+        if (min > max) throw new RuntimeException("min must be less than max");
```
Why `RuntimeException` and not an `IllegalArgumentException`?


```suggestion
        if (min > max) throw new IllegalArgumentException("min must be less than max");
```

TobiGr commented on extractor/src/main/java/org/schabi/newpipe/extractor/downloader/Response.java:
```diff
@@ -80,4 +84,129 @@ public String getHeader(final String name) {
+    /**
+     * Throw exception if response code is a 5xx server error
+     * @return this {@code Response}
+     * @throws HttpResponseException if the response code is 4xx
```
```suggestion
     * @throws HttpResponseException if the response code is 5xx
```

TobiGr reviewed (approved):
THank you. Can be merged once these two small things are applied

absurdlylongusername committed:
Fix typo
Use IllegalArgumentException instead of Runtime
Add bounds checking to all range methods
Add doc to another method

absurdlylongusername committed:
Merge branch 'dev' into soundcloud-validate-http-response

TobiGr commented:
@Stypox Do you want to review again or can we merge the PR? Your comments have been addressed.
